### PR TITLE
Feature/add work package attachments (WP 29203)

### PR DESCRIPTION
**Pull request by virgo47**
I'd like to contribute the implementation of [WP 29203](https://community.openproject.org/projects/openproject/work_packages/29203/activity), feature "Allow attachment upload on read-only work packages". The implementation is based on a new permission `:add_work_package_attachments` which works on its own or can be combined with adding notes (comments) for drag'n'drop functionality - this works as expected automatically when the attachment links are provided in the API responses.

I managed to extend one test only `spec/features/work_packages/attachments/attachment_upload_spec.rb` - and was lucky enough to conflict with incoming changes in the process. ;-)

Guys, please, give me some hints what other tests I should extend or what new tests (ideally similar to an existing ones) I should add, because I'm quite struggling to orient myself in the tests.

As for manual validation, the process is easy:
* If the member has Edit WP permission, everything works as before.
* If the member does not have Edit WP nor the new permission, everything works as before (attachments can't be added to an existing package).
* If the member has the new permission and no Edit WP (e.g. some "Attachment adder" role) then the user can add attachments to an existing WP:
  * They can d'n'd to Files area.
  * They can use the Attach button there to add it via OS open file dialog.
  * IF the role has also permission to add comments ("notes"), then they can also d'n'd into the new comment.

The new permission does not provide delete attachment capability.

### CODE REVIEW

**Comment by virgo47** [comment]
I'm on those `./spec/requests/api/v3/attachments/work_package_spec.rb` failures.

**Comment by oliverguenther** [comment]
Hi @virgo47 , thanks for the contribution!

File: config/initializers/permissions.rb [hunk 1]
```diff
@@ -309,6 +309,12 @@
                      require: :member,
                      contract_actions: { work_packages: %i[assigned] },
                      grant_to_admin: false
+
+      # WP attachments can be added with :edit_work_packages, this permission allows it without Edit WP as well.
+      wpt.permission :add_work_package_attachments,
+                     {},
+                     permissible_on: %i[work_package project],
+                     dependencies: :view_work_packages
```
Comment (cbliard) [inline config/initializers/permissions.rb]: Can you please move this permission definition after the `:edit_own_work_package_notes` so it is grouped with other similar permissions.

File: config/locales/en.yml [hunk 2]
```diff
@@ -2475,6 +2475,8 @@ en:
   permission_add_work_packages: "Add work packages"
   permission_add_messages: "Post messages"
   permission_add_project: "Create projects"
+  permission_add_work_package_attachments: "Add attachments"
+  permission_add_work_package_attachments_explanation: "Allows adding attachments without Edit work package permission"
```
Comment (cbliard) [inline config/locales/en.yml]: There is a missing s at the end of "Edit work package**s**".

**Review by cbliard — CHANGES_REQUESTED** [review]
Thank you for your contribution!

The functional tests you added cover the parts you described. If you want to also cover the "The new permission does not provide delete attachment capability." part, you could add a test `spec/requests/api/v3/work_packages/delete_resource_spec.rb` to ensure having `:add_work_package_attachments` is not enough to delete a work package, but I'm not sure if that's really needed.

I also have some minor enhancements requests about your code. I'll merge the PR once you fix them.

**Comment by virgo47** [comment]
Thanks for the review! I've made the suggested changes and also added the permission to the negative scenario in `.../work_packages/delete_resource_spec.rb`, and in the same manner for `update_resource_spec` as well. I hope you don't mind using two permissions in the test at once, but I believe the test idea stayed the same. Repeating it again in a different context didn't make sense to me - but let me know what you think. ;-)
I also merged recent dev into the branch.

**Review by cbliard — APPROVED** [review]

**Comment by cbliard** [comment]
> I hope you don't mind using two permissions in the test at once, but I believe the test idea stayed the same. Repeating it again in a different context didn't make sense to me

You're right.

Thanks a lot for implementing this. You pull request was well tested and was a pleasure to review :heart:
This feature will be released in version 13.1.

**Pull request merged**